### PR TITLE
run_function return type specificity

### PR DESCRIPTION
**Pull request by thundergolfer**
## Describe your changes

**context:** https://modal-com.slack.com/archives/C06UKEPF3HR/p1730314009504369

- _Provide Linear issue reference (e.g. MOD-1234) if available._

<details> <summary>Backward/forward compatibility checks</summary>

Just a type-checking change and docs change.

---

Check these boxes or delete any item (or this section) if not relevant for this PR.

- [x] Client+Server: this change is compatible with old servers
- [x] Client forward compatibility: this change ensures client can accept data intended for later versions of itself

### CODE REVIEW

File: modal/image.py [hunk 1]
```diff
@@ -1396,7 +1396,7 @@ def build_dockerfile(version: ImageBuilderVersion) -> DockerfileSpec:
 
     def run_function(
         self,
-        raw_f: Callable[..., Any],
+        raw_f: Callable[..., None],
```
Comment (thundergolfer) [inline modal/image.py:1399]: This is a narrowing but I can't think of a case where returning something is desired. These should just be side-effect functions.

**Comment by mwaskom** [comment]
I could imagine users doing something like this

```python
def load_model_with_caching():
    return transformers.AutoModel.from_pretrained("whatever-bert")


image = modal.Image.debian_slim().run_function(load_model_with_caching)

...

@app.function()
def do_inference(input):
    model = load_model_with_caching()
    model(input)
```

Similar to our advertised "stack `@modal.build` / `@modal.enter` pattern, to take advantage of huggingface's caching during the Image build.

Still it's helpful to clarify that we don't use the return value. Maybe that could just be done with narrative documentation? Not sure.

**Comment by thundergolfer** [comment]
Yeh I think I've even done something similar myself. But the problem is that in the `run_function` case Modal's client will actually try pickle that return value and send it as an output, which is undesirable.

**Comment by mwaskom** [comment]
> But the problem is that in the run_function case Modal's client will actually try pickle that return value and send it as an output, which is undesirable.

I guess it wouldn't be trivial not to do this since the container doesn't know it's running a "build function"?

**Comment by thundergolfer** [comment]
> I guess it wouldn't be trivial not to do this since the container doesn't know it's running a 

It might actually be easy enough. I think the container can access the Function definition which has something like `is_build_function`

**Comment by freider** [comment]
Good point, if it's not too hard structurally in the container entrypoint we should add something that just returns None as the result value when running a builder function! I've ran into this issue where builds take ages because of doing things like `run_function(load_model_with_caching)` like above to reuse the same model loading function without declaring a separate builder function

**Comment by mwaskom** [comment]
Going to close here and track the suggested change to container entrypoint in https://linear.app/modal-labs/issue/CLI-265/dont-try-to-serialize-builder-function-outputs

**Pull request closed without merging**